src/inventorymanager.cpp

```diff
@@ -471,69 +473,72 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
-		}
-		// Source is detached
-		if (from_inv.type == InventoryLocation::DETACHED) {
-			PLAYER_TO_SA(player)->detached_inventory_OnTake(
-				*this, src_item, player);
-		}
-	}
-
-	/* Node metadata inventories */
-
-	// Both endpoints are same nodemeta
-	if (from_inv.type == InventoryLocation::NODEMETA &&
-			from_inv == to_inv) {
-		PLAYER_TO_SA(player)->nodemeta_inventory_OnMove(
-			*this, count, player);
-	} else {
-		// Destination is nodemeta
-		if (to_inv.type == InventoryLocation::NODEMETA) {
+		else if (to_inv.type == InventoryLocation::NODEMETA)
 			PLAYER_TO_SA(player)->nodemeta_inventory_OnPut(
 				*this, src_item, player);
-		}
-		// Source is nodemeta
-		if (from_inv.type == InventoryLocation::NODEMETA) {
-			PLAYER_TO_SA(player)->nodemeta_inventory_OnTake(
-				*this, src_item, player);
-		}
-	}
-
-	// Player inventories
-
-	// Both endpoints are same player inventory
-	if (from_inv.type == InventoryLocation::PLAYER &&
-			from_inv == to_inv) {
-		PLAYER_TO_SA(player)->player_inventory_OnMove(
-			*this, count, player);
-	} else {
-		// Destination is player inventory
-		if (to_inv.type == InventoryLocation::PLAYER) {
+		else if (to_inv.type == InventoryLocation::PLAYER)
 			PLAYER_TO_SA(player)->player_inventory_OnPut(
 				*this, src_item, player);
-		}
-		// Source is player inventory
-		if (from_inv.type == InventoryLocation::PLAYER) {
+		// Sources
+		if (from_inv.type == InventoryLocation::DETACHED)
+			PLAYER_TO_SA(player)->detached_inventory_OnTake(
+				*this, src_item, player);
+		else if (from_inv.type == InventoryLocation::NODEMETA)
+			PLAYER_TO_SA(player)->nodemeta_inventory_OnTake(
+				*this, src_item, player);
+		else if (from_inv.type == InventoryLocation::PLAYER)
 			PLAYER_TO_SA(player)->player_inventory_OnTake(
 				*this, src_item, player);
+		if (did_swap) {
+			// Item is now placed in source list
+			src_item = list_from->getItem(from_i);
+			std::swap(from_inv, to_inv);
+			std::swap(from_list, to_list);
+			std::swap(from_i, to_i);
+			// Destinations
+			if (to_inv.type == InventoryLocation::DETACHED)
+				PLAYER_TO_SA(player)->detached_inventory_OnPut(
+					*this, src_item, player);
+			else if (to_inv.type == InventoryLocation::NODEMETA)
+				PLAYER_TO_SA(player)->nodemeta_inventory_OnPut(
+					*this, src_item, player);
+			else if (to_inv.type == InventoryLocation::PLAYER)
+				PLAYER_TO_SA(player)->player_inventory_OnPut(
+					*this, src_item, player);
+			// Sources
+			if (from_inv.type == InventoryLocation::DETACHED)
+				PLAYER_TO_SA(player)->detached_inventory_OnTake(
+					*this, src_item, player);
+			else if (from_inv.type == InventoryLocation::NODEMETA)
+				PLAYER_TO_SA(player)->nodemeta_inventory_OnTake(
+					*this, src_item, player);
+			else if (from_inv.type == InventoryLocation::PLAYER)
+				PLAYER_TO_SA(player)->player_inventory_OnTake(
+					*this, src_item, player);
+			std::swap(from_inv, to_inv);
+			std::swap(from_list, to_list);
+			std::swap(from_i, to_i);
```

Comment:
@appgurueu Would you be able to elaborate on the reason behind all these `std::swap` calls?

Reply:
@ClobberXD pretty simple actually. The `allow_*` and `on_*` callbacks are called with the MoveAction (`*this`) as argument. When swapping, we basically have two move actions. This temporarily emulates a "swapped" MoveAction - swapping source & destination. Then the callbacks are called again.

BTW technically `std::swap` is not a function but a macro (if it was a function I'd have to pass pointers), so no "calls" here ;)

Reply:
`std::swap` is a [template function](https://en.cppreference.com/w/cpp/algorithm/swap). Unlike C you can pass references without explicitly using pointers.

Reply:
@appgurueu Ah I see, thanks for clarifying.